File: Lab/src/agent-creator/AgentCreator/A2AAuth.cs
namespace AgentCreator;

public class A2AAuthFilter : IEndpointFilter
{
    private readonly bool _enabled;
    private readonly string? _token;

    public A2AAuthFilter()
    {
        var enabledVar = Environment.GetEnvironmentVariable("A2A_AUTH_ENABLED");
        _enabled = string.Equals(enabledVar, "true", StringComparison.OrdinalIgnoreCase);
        _token = Environment.GetEnvironmentVariable("A2A_AUTH_TOKEN");

        // Fail-fast: if auth enabled but no token, crash on startup
        if (_enabled && string.IsNullOrWhiteSpace(_token))
        {
            throw new InvalidOperationException("A2A_AUTH_ENABLED=true but A2A_AUTH_TOKEN is empty. Set the token or disable auth.");
        }
    }

    public async ValueTask<object?> InvokeAsync(EndpointFilterInvocationContext context, EndpointFilterDelegate next)
    {
        // Auth disabled? Pass through
        if (!_enabled)
        {
            return await next(context);
        }

        var httpContext = context.HttpContext;
        var authHeader = httpContext.Request.Headers.Authorization.FirstOrDefault();
        var apiKeyHeader = httpContext.Request.Headers["X-API-Key"].FirstOrDefault();

        string? providedToken = null;

        // Check Bearer token
        if (!string.IsNullOrEmpty(authHeader) && authHeader.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
        {
            providedToken = authHeader["Bearer ".Length..].Trim();
        }
        // Check API Key
        else if (!string.IsNullOrEmpty(apiKeyHeader))
        {
            providedToken = apiKeyHeader.Trim();
        }

        // Validate
        if (string.IsNullOrEmpty(providedToken) || providedToken != _token)
        {
            httpContext.Response.StatusCode = 401;
            return Results.Json(new { error = "Unauthorized", message = "Invalid or missing authentication token" });
        }

        return await next(context);
    }
}

File: Lab/src/agent-research/a2a_auth.py
"""A2A authentication middleware — bearer token + API key validation."""

import os
from fastapi import HTTPException, Header
from typing import Annotated

A2A_AUTH_ENABLED = os.getenv("A2A_AUTH_ENABLED", "false").lower() == "true"
A2A_AUTH_TOKEN = os.getenv("A2A_AUTH_TOKEN")

# Fail-fast: if auth is enabled but no token configured, raise on import
if A2A_AUTH_ENABLED and not A2A_AUTH_TOKEN:
    raise RuntimeError("A2A_AUTH_ENABLED=true but A2A_AUTH_TOKEN is not set")


async def verify_a2a_token(
    authorization: Annotated[str | None, Header()] = None,
    x_api_key: Annotated[str | None, Header()] = None,
):
    """Validate Bearer token or X-API-Key header. Raises 401 if auth enabled and invalid."""
    if not A2A_AUTH_ENABLED:
        return

    token = None
    if authorization and authorization.startswith("Bearer "):
        token = authorization[7:]
    elif x_api_key:
        token = x_api_key

    if token != A2A_AUTH_TOKEN:
        raise HTTPException(status_code=401, detail="Unauthorized")
